File: dcp-solutions/2024-12-03/solution.php
```php
<?php
function numIslands($matrix) {
    if (empty($matrix)) {
        return 0;
    }

    $rows = count($matrix);
    $cols = count($matrix[0]);
    $island_count = 0;

    for ($r = 0; $r < $rows; $r++) {
        for ($c = 0; $c < $cols; $c++) {
            if ($matrix[$r][$c] == '1') {
                dfs($matrix, $r, $c);
                $island_count++;
            }
        }
    }

    return $island_count;
}

function dfs(&$matrix, $r, $c) {
    if ($r < 0 || $c < 0 || $r >= count($matrix) || $c >= count($matrix[0]) || $matrix[$r][$c] == '0') {
        return;
    }

    $matrix[$r][$c] = '0';

    dfs($matrix, $r - 1, $c);
    dfs($matrix, $r + 1, $c);
    dfs($matrix, $r, $c - 1);
    dfs($matrix, $r, $c + 1);
}
```

File: dcp-solutions/2024-12-03/solution.rb
```ruby
def numIslands(grid)
  return 0 if grid.empty?

  rows, cols = grid.size, grid[0].size
  visited = Array.new(rows) { Array.new(cols, false) }

  def dfs(r, c)
    return unless (r < 0 || r >= rows || c < 0 || c >= cols || visited[r][c] || grid[r][c] == '0')
    visited[r][c] = true
    dfs(r - 1, c)
    dfs(r + 1, c)
    dfs(r, c - 1)
    dfs(r, c + 1)
  end

  (0...rows).each do |r|
    (0...cols).each do |c|
      return count = 1 if grid[r][c] == '1' && !visited[r][c]
      dfs(r, c) unless grid[r][c] == '0'
      visited[r][c] = true
      count += 1 if grid[r][c] == '1'
    end
  end

  count
end
```